DESIGN_DOC.md

```diff
@@ -0,0 +1,99 @@
+- User's login (required)
+- User's password (required)
+- Configuration parameters for the API (optional) - will be described later.
+Upon receiving the request, the API queries the Google Calendar API to get the access to the user's calendars [1] (an example is at [7]).
+
+Google supports only the OAuth protocol. This complicates the implementation:
+1. To get the API token from Google, the user must open a browser and confirm the request.
+2. When the request is confirmed, Google sends a new request to the callback endpoint containing the API key. The callback endpoint
+must be specified in the original request.
+Detailed explanation of how authorization works with Google can be found at [6].
+
+There is a Google Java client but the implementation is blocking. Also, the implementation is somewhat unusual and I think it might be hard to integrate it with
+our backend - needs to be investigated.
+
+### Calendar monitoring
+Google Calendar API supports push notifications using a webhook [2]. The only constraint is that the Google can only send requests to verified
+domain names only on HTTPS protocol. The detailed explanation of how domain names can be verified is available at [2].
+
+To receive notifications about created or updated events in the user's calendar we can use the `watch` endpoint of the Calendar API [3].
+We would need to set up an appropriate hook on our side to recieve these notifications. Our hook will recieve notifications for both newly created events
+and updated events - this will affect our preprocessing of these events which will be described later.
+
+### Disaster monitoring
+Disaster events can be queried from some provided APIs [4] [5]. Which API we use is not particularly relevant at the moment - the only difference will be in the way we determine which disaster events affect which calendar events (more on that later). We can set up a cron job that will query the disaster API on a regular basis and store received events into the database.
+
+### Matching disasters with calendar events
+A reasonable way to determine whether some disaster event affects any of the calendar events would be to select all disaster events at a specified radius from the calendar event. This method is appropriate for the [4] disaster API, which provides precise coordinates of the former.
+
+A different approach would be to match a general location of both the disaster event and the calendar event. For example, we can consider some calendar event to be affected by the disaster if both events occur in the same country. This method is suitable for the [5] API.
+
+If the first method is selected, we will need to determine the precise coordinates of the calendar event since the calendar API only provides a textual representation of the address. We can use Google's Maps API to achieve that. We can also use this API to determine the distance between two points on the globe.
```

Comment:
The coordinates of location in text could be determined during preprocessing phase (after receiving notification), and then stored along with other data in disaster events table.

Reply:
Distance between two points is easier done in code (SQL, Java), instead of network request for each such computation.

Reply:
Agree with the first comment. Regarding the second one - network computation is not required. Maps API just has some utility functions that implement the algorithm without any network requests. But we still need Maps API, though, to compute coordinates out of the address, and that requires network connection (I think).